lib/paypal.rb

```diff
@@ -0,0 +1,57 @@
+# With the existing configuration of payments on the backend,
+# it was much lower of a lift to use the existing database columns
+# that are prefixed with stripe. There are still ways to differentiate
+# subscription platforms based on the unique stripe_plan_id. With more
+# time I would restructure the database, removing the stripe_ prefix
+# and then adding an external_merchant:string column to easily tell
+# the difference in subscriptions.
+
+class Paypal
+  class Subscribe
+    if Rails.env.production?
+      UNSUB_BASE_URL = "https://api-m.paypal.com/v1/billing/subscriptions"
+    else
+      UNSUB_BASE_URL = "https://api-m.sandbox.paypal.com/v1/billing/subscriptions"
+    end
+    AUTH = { username: ENV["PAYPAL_CLIENT_ID"], password: ENV["PAYPAL_SECRET_KEY"] }
+
+    # create(data: hash, user: User)
+    # Finds all paypal subscriptions and cancels them
+    # Creates a subscription for the user based on data returned from paypal.
+    # This data is returned via the "Auto Return URL" specified within Account Settings >
+    # Website Payments > Website Preferences. This url should point to /api/webhooks/paypal.
+    def self.create(data, user)
+      cancel_all(user)
+
+      order_id = data["orderID"]
+      billing_token = data["billingToken"]
+      subscription_id = data["subscriptionID"]
+
+      user.subscriptions.create!(
+        stripe_plan_id: ENV["PAYPAL_PLAN_ID"],
+        stripe_subscription_id: subscription_id,
+        stripe_session_id: order_id,
+        stripe_payment_intent_id: billing_token
+      )
+    end
```

Comment:
When someone tries that they will get a `500` since there is no user. It might be better if I throw in a hook to ensure the user exists before either of these methods.

Reply:
If it's done without passing any cookies then sure `current_user` will be nil, but a user could pass their own cookies and it would bypass that.

Steps to reproduce:
1. Open [beta.splits.io](https://beta.splits.io)
2. Make sure you are logged in
3. Open a JavaScript console in the page
4. Type this:
    ```javascript
    fetch('/api/webhooks/paypal', {method: 'POST'})
    ```
Free subscription 😛

Reply:
Back to the drawing board on this one then. Would be nice if PayPal figured out their v2 API for subscriptions. :(

Reply:
I'm not super familiar with PayPal, but it may be as simple as checking with the PayPal API to see if the order ID (or similar) really exists. And we can validate its uniqueness on our end to make sure they're not maliciously reused.

Reply:
That sounds like a good solution. Thanks for the tip!